tests/testthat/test_dump.R

```diff
@@ -0,0 +1,270 @@
+  expect_true(file.exists(f))
+  data <- read.csv(f, stringsAsFactors = FALSE)
+  expect_equal(nrow(data), 1)
+  expect_equal(data$modelling_group, "LAP-elf")
+  expect_equal(data$disease, "flu")
+  expect_equal(data$scenario, "pies")
+  expect_equal(data$scenario_type, "standard")
+  expect_equal(data$age_min_inclusive, 0)
+  expect_equal(data$age_max_inclusive, 85)
+  expect_equal(data$year_min_inclusive, 2000)
+  expect_equal(data$year_max_inclusive, 2100)
+  expect_equal(data$cohort_min_inclusive, 1915)
+  expect_equal(data$cohort_max_inclusive, 2100)
+  expect_equal(data$countries, "AFG;ZWE")
+  expect_equal(data$outcomes, "cases;deaths")
+  expect_equal(data$touchstone, "nevis-1")
+}
+
+test_dump_db_modelling_group <- function(path) {
+  f <- file.path(path, "db_modelling_group.csv")
+  expect_true(file.exists(f))
+  data <- read.csv(f, stringsAsFactors = FALSE)
+  expect_equal(nrow(data), 1)
+  expect_equal(data$id, "LAP-elf")
+  expect_equal(data$institution, "Lapland Epi Centre")
+  expect_equal(data$pi, "Santa")
+  expect_equal(data$description, "Description")
+  expect_true(is.na(data$comment))
+  expect_true(is.na(data$replaced_by))
+}
+
+test_dump_db_disease <- function(path, id = "flu", name = "Elf") {
+  f <- file.path(path, "db_disease.csv")
+  expect_true(file.exists(f))
+  data <- read.csv(f, stringsAsFactors = FALSE)
+  expect_equal(nrow(data), 2)
+  expect_equal(data$id, "flu")
+  expect_equal(data$name, "Elf flu")
+}
+
+test_dump_db_disease_extra <- function(path) {
+  f <- file.path(path, "db_disease.csv")
+  expect_true(file.exists(f))
+  data <- read.csv(f, stringsAsFactors = FALSE)
+  expect_equal(nrow(data), 2)
+  expect_equal(data$id, c("flu", "piles"))
+  expect_equal(data$name, c("Elf flu", "Elf piles"))
+}
+
+test_that("Dump non-existent Touchstone", {
+  test <- new_test()
+  tmp <- empty_dump()
+  expect_error(stoner::stone_dump(test$con, "turnip", tmp, TRUE),
+               "Touchstone not found - ")
+})
+
+test_that("Dump works (touchstone only)", {
+  test <- initialise_touchstones(new_test())
+  tmp <- empty_dump()
+  stoner::stone_dump(test$con, "nevis-1", tmp, TRUE)
+  test_dump_touchstone(tmp)
+  test_dump_touchstone_name(tmp)
+  expect_equal(length(list.files(tmp)), 2)
+})
+
+test_that("Dump works (touchstone + demography)", {
+  test <- initialise_with_demography(new_test())
+  tmp <- empty_dump()
+  stoner::stone_dump(test$con, "nevis-1", tmp, FALSE)
+  test_dump_touchstone(tmp)
+  test_dump_touchstone_name(tmp)
+  test_dump_touchstone_dds(tmp)
+  expect_equal(length(list.files(tmp)), 3)
+  stoner::stone_dump(test$con, "nevis-1", tmp, TRUE)
+  test_dump_touchstone(tmp)
+  test_dump_touchstone_name(tmp)
+  test_dump_touchstone_dds(tmp)
+  test_dump_db_demographic_source(tmp)
+  test_dump_db_demographic_statistic_type(tmp)
+  expect_equal(length(list.files(tmp)), 5)
+})
+
+# Re-implement below when expections support is done
+
+test_that("Dump works (fake expectations)", {
```

Comment:
I'm a bit confused about this testing format, but I think I can see what you're doing. More a comment than anything

Reply:
I've added a bunch of comments into the test, as I will probably forget why I did this.

Essentially, fake data is pushed into the db by a number of functions in helper-tests.R, and the functions right at the top. I then have a number of functions that test that a specified CSV (of which stoner dump can make about 10) matches the fake data that I've put in one way or another - these are called from multiple tests.

Lastly, the actual tests are kind of incremental, to test the combinations of how many things might be linked to a touchstone: 

(1) NULL case - touchstone doesn't exist. 
(2) the touchstone exists - and must have matching touchstone_name 
(3) additionally, a touchstone_demographic_dataset links to that touchstone.
(4) additionally, a responsibility_set (without responsibilities) links to that touchstone.
(5) additionally, responsibilities and expectations exist (they come together) - but no countries or outcomes
(6...) finally, add in combinations of and/or countries/outcomes.

dump should work correctly and produce the right files in all of those in/complete situations, handling any absences appropriately. I also test for each the "include_deps" T/F flag, which determines whether dependent data (modelling group, demographic source/type) is dumped as db_...csv files.

Reply:
(It is probably a bit overkill - although as the number of tests on different things went up, a bug popped out - in the fresh db, serial ids often start at 1, not to be confused with the first index of a vector... 

It might be good if in test_db scenarios, the serial numbers of any id fields are bumped to some random number at the start!)